core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
+    private void copyBytes(Slice rawSlice, int[] lengths, int[] sourceOffsets, int count, int newByteCount)
     {
         ensureBytesCapacity(getCurrentOffset() + newByteCount);
 
-        for (int i = 0; i < count; i++) {
-            int position = positions[i];
-            if (!block.isNull(position)) {
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
+            for (int i = 0; i < count; i++) {
                 int length = lengths[i];
-                Slice slice = block.getSlice(position, 0, length);
-                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+                if (length > 0) {
+                    int offset = sourceOffsets[i];
+                    System.arraycopy(base, offset, bytes, offsets[positionCount + i], length);
+                }
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                int length = lengths[i];
+                if (length > 0) {
+                    int offset = sourceOffsets[i];
+                    rawSlice.getBytes(offset, bytes, offsets[positionCount + i], length);
+                }
             }
         }
-
         positionCount += count;
         updateSize(count, newByteCount);
     }
 
     /**
-     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     * Copy {@code length} bytes from {@code rawSlice}, starting at offset {@code sourceOffset} to {@code count} consecutive positions in the {@link #bytes} array.
      */
-    private void duplicateBytes(Block block, int position, int count)
+    private void duplicateBytes(Slice rawSlice, int sourceOffset, int length, int count)
     {
-        int length = block.getSliceLength(position);
         int newByteCount = toIntExact((long) count * length);
         int startOffset = getCurrentOffset();
         ensureBytesCapacity(startOffset + newByteCount);
 
-        Slice slice = block.getSlice(position, 0, length);
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
+            for (int i = 0; i < count; i++) {
+                System.arraycopy(base, sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                rawSlice.getBytes(sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
         for (int i = 0; i < count; i++) {
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
-            offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
+            this.offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
```

Comment:
nit: Could be simplified to something like
```
int currentStartOffset = startOffset + length;
for (int i = 0; i < count; i++) {
    offsets[positionCount + i + 1] = currentStartOffset;
    currentStartOffset += length;
}
```

Reply:
I missed this comment.
applied now.